kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,273 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+    for (auto public_event_group_id : public_event_group_ids) {
+      eg_state->event_group_id_batch.emplace_back(public_event_group_id);
+    }
+    ConcordAssertLE(eg_state->event_group_id_batch.size(), kBatchSize);
+    // increment the offset
+    eg_state->public_offset += eg_state->event_group_id_batch.size();
+    LOG_DEBUG(logger_,
+              "Updated public_offset: " << eg_state->public_offset
+                                        << " public_event_group_ids size: " << public_event_group_ids.size());
+  } else if (!private_event_group_ids.empty() &&
+             (public_event_group_ids.empty() || (private_event_group_ids.back() <= public_event_group_ids.front()))) {
+    for (auto private_event_group_id : private_event_group_ids) {
+      eg_state->event_group_id_batch.emplace_back(private_event_group_id);
+    }
+    ConcordAssertLE(eg_state->event_group_id_batch.size(), kBatchSize);
+    // increment the offset
+    eg_state->private_offset += eg_state->event_group_id_batch.size();
+    LOG_DEBUG(logger_,
+              "Updated private_offset: " << eg_state->private_offset
+                                         << " private_event_group_ids size: " << private_event_group_ids.size());
+  } else if (!public_event_group_ids.empty() && !private_event_group_ids.empty()) {
+    LOG_DEBUG(logger_, "Both public and private event groups found in the batch");
+    std::merge(public_event_group_ids.begin(),
+               public_event_group_ids.end(),
+               private_event_group_ids.begin(),
+               private_event_group_ids.end(),
+               std::back_inserter(eg_state->event_group_id_batch));
+    // We cannot return the full merge
+    // E.g.: a=[1,2,3,4,5] b=[6,7,8,9]
+    if (eg_state->event_group_id_batch.size() > kBatchSize) {
+      eg_state->event_group_id_batch.erase(eg_state->event_group_id_batch.begin() + kBatchSize,
+                                           eg_state->event_group_id_batch.end());
+    }
+    // find and update the offsets \
+    // We iterate over event_group_id_batch in the reverse order, and update the private/public offset if the event group id
+    // in event_group_id_batch is found in private_event_group_ids/public_event_group_ids respectively.\
+    // In the worst case, we need to read the entire event_group_id_batch (of size kBatchSize) to find the offset
+    bool has_pub_offset_updated = false;
+    bool has_pvt_offset_updated = false;
+    for (auto r_it = eg_state->event_group_id_batch.rbegin(); r_it != eg_state->event_group_id_batch.rend(); r_it++) {
+      if (has_pub_offset_updated && has_pvt_offset_updated) break;
+      if (!has_pub_offset_updated && std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it) !=
+                                         public_event_group_ids.end()) {
+        auto pub_it = std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it);
```

Comment:
Calling `find` twice here with the same parameters screams optimization :)
This is more lines of code but we call find only once (requires C++17 which is enabled):
```
if (!has_pub_offset_updated) {
  if (auto pub_it = find(...); pub_it != end()) {
    ++pub_it;
  }
}
```
Another suggestion is to implement `merge` ourselves and within track the offset.
@shgandhi what do you think?

Reply:
Yes, I agree, updated acc. to your suggestion.